modules/placement-driver/src/main/java/org/apache/ignite/internal/placementdriver/leases/LeaseTracker.java

```diff
@@ -257,4 +247,46 @@ private PendingIndependentComparableValuesTracker<HybridTimestamp, ReplicaMeta>
     ) {
         return primaryReplicaWaiters.computeIfAbsent(groupId, key -> new PendingIndependentComparableValuesTracker<>(MIN_VALUE));
     }
+
+    private CompletableFuture<Void> loadLeasesBusyAsync(long recoveryRevision) {
+        Entry entry = msManager.getLocally(PLACEMENTDRIVER_LEASES_KEY, recoveryRevision);
+
+        CompletableFuture<Void> loadLeasesFuture;
+
+        if (entry.empty() || entry.tombstone()) {
+            leases = new Leases(Map.of(), BYTE_EMPTY_ARRAY);
+
+            loadLeasesFuture = completedFuture(null);
+        } else {
+            byte[] leasesBytes = entry.value();
+
+            LeaseBatch leaseBatch = LeaseBatch.fromBytes(ByteBuffer.wrap(leasesBytes).order(LITTLE_ENDIAN));
+
+            Map<ReplicationGroupId, Lease> leasesMap = new HashMap<>();
+
+            List<CompletableFuture<?>> fireEventFutures = new ArrayList<>();
+
+            leaseBatch.leases().forEach(lease -> {
+                leasesMap.put(lease.replicationGroupId(), lease);
+
+                if (lease.isAccepted()) {
+                    getOrCreatePrimaryReplicaWaiter(lease.replicationGroupId()).update(lease.getExpirationTime(), lease);
+
+                    fireEventFutures.add(fireEventReplicaBecomePrimary(recoveryRevision, lease));
```

Comment:
Primary replica is basically a lease based entity, meaning that it's time bounded. That actually means that with PRIMARY_REPLICA_ELECTED event it'll be possible to get outdated PrimaryReplica especially within node restart case. So it should be explicitly stated within javadoc that it's possible to get stale primary and it's required to check whether the lease is outdated while trying to perform an action.